salt/crypt.py

```diff
@@ -149,41 +171,63 @@ def gen_keys(keydir, keyname, keysize, user=None, passphrase=None):
+    timestamp of the file's last modification. This fn is memoized so if it is
+    called with the same path and timestamp (the file's last modified time) the
+    second time the result is returned from the memoiziation.  If the file gets
+    modified then the params are different and the key is loaded from disk.
     '''
     log.debug('salt.crypt._get_key_with_evict: Loading private key')
-    with salt.utils.files.fopen(path) as f:
-        key = RSA.importKey(f.read(), passphrase)
+    if HAS_M2:
+        key = RSA.load_key(path, lambda x: six.b(passphrase))
+    else:
+        with salt.utils.files.fopen(path) as f:
+            key = RSA.importKey(f.read(), passphrase)
     return key
 
 
-def _get_rsa_key(path, passphrase):
+def get_rsa_key(path, passphrase):
     '''
-    Read a key off the disk.  Poor man's simple cache in effect here,
-    we memoize the result of calling _get_rsa_with_evict.  This means
-    the first time _get_key_with_evict is called with a path and a timestamp
-    the result is cached.  If the file (the private key) does not change
-    then its timestamp will not change and the next time the result is returned
-    from the cache.  If the key DOES change the next time _get_rsa_with_evict
-    is called it is called with different parameters and the fn is run fully to
-    retrieve the key from disk.
+    Read a private key off the disk.  Poor man's simple cache in effect here,
+    we memoize the result of calling _get_rsa_with_evict.  This means the first
+    time _get_key_with_evict is called with a path and a timestamp the result
+    is cached.  If the file (the private key) does not change then its
+    timestamp will not change and the next time the result is returned from the
+    cache.  If the key DOES change the next time _get_rsa_with_evict is called
+    it is called with different parameters and the fn is run fully to retrieve
+    the key from disk.
     '''
-    log.debug('salt.crypt._get_rsa_key: Loading private key')
+    log.debug('salt.crypt.get_rsa_key: Loading private key')
     return _get_key_with_evict(path, six.text_type(os.path.getmtime(path)), passphrase)
 
 
+def get_rsa_pub_key(path):
+    '''
+    Read a public key off the disk.
+    '''
+    log.debug('salt.crypt.get_rsa_pub_key: Loading public key')
+    if HAS_M2:
+        key = RSA.load_pub_key(path)
+    else:
+        with salt.utils.files.fopen(path) as f:
+            key = RSA.importKey(f.read())
+    return key
+
+
 def sign_message(privkey_path, message, passphrase=None):
     '''
     Use Crypto.Signature.PKCS1_v1_5 to sign a message. Returns the signature.
     '''
-    key = _get_rsa_key(privkey_path, passphrase)
+    key = get_rsa_key(privkey_path, passphrase)
     log.debug('salt.crypt.sign_message: Signing message.')
-    signer = PKCS1_v1_5.new(key)
-    return signer.sign(SHA.new(message))
+    if HAS_M2:
+        md = EVP.MessageDigest('sha1')
```

Comment:
sha1? This brings to mind the 2017 collision attacks on sha1.

cc: @thatch45

Reply:
Crypto in salt.crypt uses SHA hash package that does SHA1 signature.
Here is the same as we discussed above with Tom: if we'd like to use for instance SHA256 we have to either change this behavior in both M2Crypto and pyCrypto implementations or we'll break master-minion compatibility if they run in different configurations.
What I've done in this PR is cloned the current Crypto behavior replaced it with M2Crypto to keep the compatibility.
So I propose to keep this PR as is and do any feature improvements in different issues. And the first one we'd have to do is modularize the crypt package. This'd make it configurable and testable including writing compatibility tests.

Reply:
But anyway if you'd like to have sha256 in the oxygen I can do the change.

Reply:
That is fine. 👍

Reply:
Thank you!